docs/internals/constants.md

```diff
@@ -0,0 +1,147 @@
+
+All the lowered HLO functions will take one additional argument
+for each unique constant appearing in the `Jaxpr` to which it corresponds.
+These arguments, referred to as `const_args`,
+come after the dimension variable arguments, after the
+token arguments, and just before the actual array arguments.
+During lowering we maintain a mapping `const_lowering: dict[int, mlir.IrValues]`
+from the `id` of the constants to the HLO values for the corresponding
+const args.
+This mapping is stored in the `mlir.LoweringRuleContext`.
+
+When we lower an HLO inner function (i.e., not the `main` function),
+we call again `core.jaxpr_const_args`
+to get the actual constants in the corresponding `Jaxpr`. These are
+expected to be among the constants for which we have a `const_lowering`.
+The inner function will get its own smaller set of `const_args` and
+its own `const_lowering` mapping to be used when lowering the body.
+E.g., the function `mlir.lower_jaxpr_as_fun` is one place where some
+of this happens.
+
+The function `mlir.jaxpr_subcomp` does not create a new HLO function,
+but instead creates a block within the current function. It uses
+the enclosing function's `const_lowering`.
+
+Note also that there will still be `stablehlo.constant` in the lowered
+code, in three cases:
+  * when the constant is a scalar; we want these constants to be
+  available to XLA for constant folding.
+  * when the constant did not appear in the traced program, and is
+  hence not in the `Jaxpr`. This can happen for constants that
+  arise during lowering, e.g., the lowering of some PRNG functions
+  include constants.
+  * when we are exporting: at the moment, we do not hoist constant args
+  when we export because the export serialization does not currently support
+  serialization of arrays.
+  We use the `mlir.LoweringParameters.hoist_constants_as_args` parameter
+  to control this.
+
+One additional complication is that some of the internal lowering functions
+need to take the argument avals and sometimes also the shardings and
+layouts for the arguments. Furthermore, the avals, shardings, and layout for
+all arguments, including the const args,
+are used also after lowering also. Therefore, it is convenient
+to compute these fairly high in the call stack, e.g., in
+`pxla.lower_sharding_computations`, and pass them down.
+
+For example, the functions `mlir.lower_jaxpr_to_module`,
+`pjit._pjit_cached_lower_jaxpr_to_fun`, and, `mlir.lower_jaxpr_to_fun`
+take `in_avals`, `in_shardings`, and `in_layouts` that
+that include both the avals for const_args and for the regular args
+(the ones corresponding to the `Jaxpr.invars`).
+They also take a `num_const_args` argument.
+
+The result of lowering, e.g., `pxla.MeshComputaton`,
+stores both a HLO module and a list
+of `const_args`. When we execute the result of the compilation
+we pass the `const_args` ahead of the actual arguments.
+We include the `const_args` in the `pxla.MeshExecutableFastpathData`
```

Comment:
This is not true yet right?

Reply:
You are right. I have removed this part, and will write it when I make some more changes to the execution path.